migrate.go

```diff
@@ -752,37 +759,52 @@ func (m *Migrate) runMigrations(ret <-chan interface{}) error {
 		case *Migration:
 			migr := r
 
-			// set version with dirty state
-			if err := m.databaseDrv.SetVersion(migr.TargetVersion, true); err != nil {
-				return err
-			}
-
 			if migr.Body != nil {
-				m.logVerbosePrintf("Read and execute %v\n", migr.LogString())
-				if err := m.databaseDrv.Run(migr.BufferedBody); err != nil {
+				// Check if the migration contains an SQL
+				// migration.
+				hasSqlMig, err := classifyMigrationBody(
+					migr, classifyPeekLimit,
+				)
+				if err != nil {
 					return err
 				}
 
-				migrVer := migr.Version
+				ver := migr.Version
+
+				// Check if the migration contains a
+				// programmatic migration.
+				_, hasPMig := m.opts.programmaticMigrations[ver]
+
+				// Execute the SQL migration or the programmatic
+				// migration.
+				switch {
+				case hasSqlMig && hasPMig:
+					return ErrMixedMigration
+
+				case hasSqlMig:
+					if err = m.databaseDrv.SetVersion(migr.TargetVersion, true); err != nil {
+						return err
+					}
 
-				// If there is a programmatic migration function
-				// for this migration, run it now.
-				cb, ok := m.opts.programmaticMigrations[migrVer]
-				if ok {
-					m.logVerbosePrintf("Running "+
-						"programmatic migration for "+
-						"%v\n", migr.LogString())
+					m.logVerbosePrintf("Read and execute %v\n", migr.LogString())
+					if err = m.databaseDrv.Run(migr.BufferedBody); err != nil {
+						return err
+					}
 
-					err := cb(migr, m.databaseDrv)
+				case hasPMig:
+					err = m.execProgrammaticMigration(migr)
 					if err != nil {
-						return fmt.Errorf("failed to "+
-							"execute programmatic "+
-							"migration: %w", err)
+						return fmt.Errorf("execution "+
+							"of programmatic "+
+							"migration failed: "+
+							"%w", err)
 					}
 
-					m.logVerbosePrintf("Programmatic"+
-						"migration finished for "+
-						"%v\n", migr.LogString())
+				default:
+					// When the migration contains no SQL
```

Comment:
hmm I wonder if we should at least log a warning that does not to seem a normal behaviour having a migration just for counting the number up.

Reply:
Added :)